analyzers/src/SonarAnalyzer.Common/DiagnosticAnalyzer/DiagnosticFactory.cs

```diff
@@ -0,0 +1,31 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2022 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace SonarAnalyzer.Helpers;
+
+public static class DiagnosticFactory
+{
+    public static Diagnostic Create(DiagnosticDescriptor descriptor, Compilation compilation, Location location, IEnumerable<Location> additionalLocations, params object[] messageArgs) =>
```

Comment:
Is this really needed? I would have expected `AnalysisContextExtensions.ReportIssue(ReportingContext reportingContext, bool? isTestProject, bool? isScannerRun)` to be changed instead. That way, the issue is gone for all "reports".

Reply:
Maybe I'm missing something but at that point, the diagnostic is already created and the class is immutable.

Reply:
You can do something like this in `sonar-dotnet\analyzers\src\SonarAnalyzer.Common\Helpers\ReportingContext.cs`
```cs
        private ReportingContext(Diagnostic diagnostic,
                                 Action<Diagnostic> contextSpecificReport,
                                 Compilation compilation,
                                 SyntaxTree syntaxTree)
        {
            SyntaxTree = syntaxTree;
            Compilation = compilation;
            Diagnostic = FixDiagnostic(diagnostic);
            this.contextSpecificReport = contextSpecificReport;
        }

        private Diagnostic FixDiagnostic(Diagnostic diagnostic)
        {
            if (diagnostic.AdditionalLocations.Any(x => !Compilation.ContainsSyntaxTree(x.SourceTree)))
            {
                return Diagnostic.Create(
                    id: diagnostic.Id,
                    category: diagnostic.Descriptor.Category,
                    message: diagnostic.Descriptor.MessageFormat,
                    severity: diagnostic.Severity,
                    defaultSeverity: diagnostic.DefaultSeverity,
                    isEnabledByDefault: diagnostic.Descriptor.IsEnabledByDefault,
                    warningLevel: diagnostic.WarningLevel,
                    isSuppressed: diagnostic.IsSuppressed,
                    title: diagnostic.Descriptor.Title,
                    description: diagnostic.Descriptor.Description,
                    helpLink: diagnostic.Descriptor.HelpLinkUri,
                    location: diagnostic.Location,
                    additionalLocations: diagnostic.AdditionalLocations.Where(x => Compilation.ContainsSyntaxTree(x.SourceTree)),
                    customTags: diagnostic.Descriptor.CustomTags,
                    properties: diagnostic.Properties);
            }
            return diagnostic;
        }
```

Reply:
There is one problem though: `messageArgs` seem to be lost in the copy.

Reply:
In this case, although it has the advantage that we don't have to change so much code, I really don't like the fact that we create a possible invalid diagnostic just to re-create it again later. I prefer to do the right thing from the start. 

Losing the `messageArgs` can be a big problem since we are using them quite a lot.

Reply:
I'm fine with both approaches. You could argue to even do both: Be careful to create the diagnostic but have a safety net for cases that slip through.
It is just too easy to forget about this problem when you create a diagnostic.

Reply:
> It is just too easy to forget about this problem when you create a diagnostic.

Indeed. I'm thinking if we shouldn't add a diagnostic to check the usages :D

Reply:
We could add some custom tag to the diagnostic in "DEBUG" and check in "ReportIssue" if the tag is set. That way, we make sure every diagnostic is created with the factory.